Ultradns rest/nppdeve 1718/migration of six rest api

devulapalli8 opened the pull request:
please review code changes done in ultradns-rest module

arghya-neustar committed:
Addition of new module named ultradns-rest

arghya-neustar committed:
Restructured ultradns-rest module

arghya-neustar committed:
Completed migration for getNeustarNetworkStatus, getResourceRecordsOfZone, getResourceRecordsOfDNameByType

sathishvc-neustar committed:
A test case `listsAllProvidersWithCredentials` was failing. Fixed it.

The test case `listsAllProvidersWithCredentials` was failing because
ultradns-rest related info was present but not expected in the return
value of `ListProviders.providerAndCredentialsTable()`. Added the
ultradns-rest related info in the expected value.

arghya-neustar committed:
NPPDEVE-1718 Migration of 6 SOAP API to REST

API migrated for this changes are getNeustarNetworkStatus, getAccountsListOfUser, getResourceRecordsOfZone, getResourceRecordsOfDNameByType, createResourceRecord, updateResourceRecord
Added test classes for changes this migrated API
Added apache common-lang dependency for string utilitiues

Kushalraj-Bhandari commented on ultradns-rest/src/main/java/denominator/ultradns/UltraDNSRestProvider.java:
```diff
@@ -0,0 +1,187 @@
+
+  private final String url;
+
+  public UltraDNSRestProvider() {
```
Do we need this default constructor for any specific reason? I would suggest to have only the parameterized constructor, unless the default one is required for some framework needs.

arghya-neustar replied:
Yes, we need that. UltraDNSRestProvider is getting constructed with all it's dependencies via dagger framework. Dagger is creating of application object and binding them together via dependency injection.

Kushalraj-Bhandari commented on ultradns-rest/src/main/java/denominator/ultradns/UltraDNSRestProvider.java:
```diff
@@ -0,0 +1,187 @@
+   */
+  public UltraDNSRestProvider(String url) {
+    this.url =
+            url == null || url.isEmpty() ? "https://test-restapi.ultradns.com/v2" : url;
```
Better configure this default url in some property file & use it here.

Kushalraj-Bhandari replied:
To make the code cleaner, I think its better to use the utility method from guava or apache for the null/empty checks.

Kushalraj-Bhandari commented on ultradns-rest/src/main/java/denominator/ultradns/UltraDNSRestProvider.java:
```diff
@@ -0,0 +1,187 @@
+  public Set<String> basicRecordTypes() {
+    Set<String> types = new LinkedHashSet<String>();
+    types.addAll(
+            Arrays.asList("A", "AAAA", "CNAME", "HINFO", "MX", "NAPTR", "NS", "PTR", "RP", "SOA", "SPF",
```
I think its better to create an enum for all these record type constants & use that enum throughout the code.

Kushalraj-Bhandari commented on ultradns-rest/src/main/java/denominator/ultradns/UltraDNSRestProvider.java:
```diff
@@ -0,0 +1,187 @@
+            new LinkedHashMap<String, Collection<String>>();
+    profileToRecordTypes.put("geo",
+            Arrays
+                    .asList("A", "AAAA", "CNAME", "HINFO", "MX", "NAPTR", "PTR", "RP",
```
Use enum instead of these constants

Kushalraj-Bhandari commented on ultradns-rest/src/main/java/denominator/ultradns/UltraDNSRestProvider.java:
```diff
@@ -0,0 +1,187 @@
+       * UltraDNSRest#addDirectionalPoolRecord(UltraDNSRest.DirectionalRecord, UltraDNSRest.DirectionalGroup, String)} can take up
+       * to 10 minutes to complete.
+       */
+      Options options = new Options(10 * 1000, 10 * 60 * 1000);
```
Better to configure this value in a property file.

Kushalraj-Bhandari commented on ultradns-rest/src/test/java/denominator/ultradns/MockUltraDNSRestServer.java:
```diff
@@ -0,0 +1,96 @@
+package denominator.ultradns;
```
Please add some more tests for all the classes. Seems like the current UT code coverage is less than 40%.

arghya-neustar replied:
We are working on the code coverage to make it greater than 80%. Once we deliver all the unit tests,  this class will be surely covered & you can see the same in jacoco reports at that time.

arghya-neustar committed:
Made modifications as per code review comments & discussion

sathishvc-neustar committed:
Added the test case testUltraDNSRestMetadata() in UltraDNSRestProviderTest

sathishvc-neustar committed:
Added the test case testUltraDNSRestRegistered() in UltraDNSRestProviderTest

arghya-neustar commented on ultradns-rest/src/main/java/denominator/ultradns/UltraDNSRestProvider.java:
```diff
@@ -71,16 +86,35 @@ public String url() {
+    ));
+
+    profileToRecordTypes.put("roundRobin", Arrays.asList(
+            ResourceTypes.A.name(),
```
` profileToRecordTypes.put(String, Collection<String>)` - is getting used only twice which is predefined. If we create a separate method to add values to a Map it won't beneficial in terms of code reuse & decrement of lines of code.
As it is has a predefined number of usage, keeping it as it is.

Kushalraj-Bhandari replied:
I disagree, what if someone, in future, wanted to update the collection & only do it at one place & misses the second one?

arghya-neustar replied:
These two collections are different. First one with key "geo" with 10 record types & second with key "roundRobin" with 11 record types. Collection<String>is not identical for "geo" & "roundRobin".
So have to update "geo" or "roundRobin" individually as their content are not same.

arghya-neustar commented on ultradns-rest/src/main/java/denominator/ultradns/model/ZoneList.java:
```diff
@@ -0,0 +1,40 @@
+
+    public List<String> getZoneNames() {
+        List<String> zoneNames = new ArrayList<String>();
+        for(Zone zone : getZones()){
```
Incorporated

sathishvc-neustar committed:
Added the test case testProviderWiresUltraDNSRestZoneApi() in UltraDNSRestProviderTest

sathishvc-neustar committed:
Added the test case testCredentialsRequired() in UltraDNSRestProviderTest

sathishvc-neustar committed:
Added the test case testTwoPartCredentialsRequired() in UltraDNSRestProviderTest

sathishvc-neustar committed:
Added the test case testViaDagger() in UltraDNSRestProviderTest

arghya-neustar commented on ultradns-rest/src/main/java/denominator/ultradns/UltraDNSRestPropertyLoader.java:
```diff
@@ -0,0 +1,31 @@
+import java.io.IOException;
+import java.util.Properties;
+
+public class UltraDNSRestPropertyLoader {
```
Incorporated changes in UltraDNSRestPropertyLoader to make it a Singleton

arghya-neustar commented on ultradns-rest/src/main/java/denominator/ultradns/InvalidatableTokenProvider.java:
```diff
@@ -37,8 +37,8 @@
         this.sessionValid = sessionValid;
-        // As per UltraDNS Authentication API Response
-        this.durationMillis = 3600 * 1000;
+        this.durationMillis = Long.parseLong(UltraDNSRestPropertyLoader
```
Addressed this issue in UltraDNSRestPropertyLoader.java

arghya-neustar commented on ultradns-rest/src/main/resources/application.properties:
```diff
@@ -0,0 +1,7 @@
+# Feign Client configuration
```
If we add this to this file i.e. application.properties other than classpath, then we have to explicitly mention the classpath while invoking denominator jar. So it will have the dependency to client which we don't want.

arghya-neustar committed:
Incorporated changes as per code review comments
Indented the Feign request body in createPrimaryZone API call

arghya-neustar committed:
Modified test configuration files
Added new unit test cases for UltraDNSRestZoneApi service, utility classes like property-loader, error-decoder

arghya-neustar committed:
Added dependency for feign-httpclient to support HTTP PATCH request

arghya-neustar committed:
Addition of jacoco plugin
